backend/src/main/java/ai/verta/modeldb/versioning/blob/visitors/Validator.java

```diff
@@ -1,20 +1,391 @@
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPathDatasetDiff;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPythonEnvironmentBlob;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPythonEnvironmentDiff;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPythonRequirementEnvironmentBlob;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenPythonRequirementEnvironmentDiff;
 import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenS3DatasetBlob;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenS3DatasetComponentBlob;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenS3DatasetComponentDiff;
+import ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.AutogenS3DatasetDiff;
+import ai.verta.modeldb.versioning.blob.diff.ProtoType;
 import io.grpc.Status;
+import io.grpc.Status.Code;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
 
 public class Validator extends Visitor {
+
+  private static final String HAS_A_STRING_VALUE_WHICH_IS_NOT_IN_A_VALID_NUMERIC_NOTATION =
+      "has a STRING_VALUE which is not in a valid numeric notation";
+
+  public Validator() {}
+
+  private boolean isNull(ProtoType proto) {
+    return proto == null;
+  }
+
+  public void validate(AutogenBlobDiff autogenBlobDiff) throws ModelDBException {
+    if (isNull(autogenBlobDiff)) {
+      return;
+    }
+    if (autogenBlobDiff.getCode() == null
+        && autogenBlobDiff.getDataset() == null
+        && autogenBlobDiff.getEnvironment() == null
+        && autogenBlobDiff.getConfig() == null) {
+      throw new ModelDBException("Unknown blob type specified", Status.Code.INVALID_ARGUMENT);
+    }
+    if (autogenBlobDiff.getLocation() == null || autogenBlobDiff.getLocation().isEmpty()) {
+      throw new ModelDBException("Blob diff location is empty", Status.Code.INVALID_ARGUMENT);
+    }
+    autogenBlobDiff.preVisitDeep(this);
+  }
+
+  public void preVisitDeepAutogenDatasetDiff(AutogenDatasetDiff blob) throws ModelDBException {
+    preVisitDeep(blob);
+  }
+
+  public void preVisitDeepAutogenS3DatasetDiff(AutogenS3DatasetDiff blob) throws ModelDBException {
+    preVisitDeep(blob);
+  }
+
+  public void validate(AutogenBlob autogenBlob) throws ModelDBException {
+    preVisitDeep(autogenBlob);
+    if (autogenBlob.getCode() == null
+        && autogenBlob.getDataset() == null
+        && autogenBlob.getEnvironment() == null
+        && autogenBlob.getConfig() == null) {
+      throw new ModelDBException("Unknown blob type specified", Status.Code.INVALID_ARGUMENT);
+    }
+  }
+
+  private void preVisitDeep(ProtoType proto) throws ModelDBException {
+    if (isNull(proto)) {
```

Comment:
Should we just fix this in the base class? Seems like an important fix that will save a lot of duplicated code too.

Reply:
I see that postVisit has already all necessary null checks. Can I use post visit instead of pre visit like in the last commit?

Reply:
Sure. It does sound like a bug on the pre, so that needs to be fixed anyway.